build/install/installer/src/win/logic.rs

```diff
@@ -0,0 +1,263 @@
+
+    Ok(())
+}
+
+/// Set application as the default handler for the given URL protocol, e.g. `enso://`.
+///
+/// This is necessary for the deep linking to work.
+pub fn register_url_protocol(executable_path: &Path, protocol: &str) -> Result {
+    let info = enso_install::win::prog_id::ProtocolInfo::new(protocol, executable_path);
+    info.register()
+}
+
+/// Register the uninstaller in the Windows registry.
+pub fn register_uninstaller(
+    config: &Config,
+    install_directory: &Path,
+    uninstaller_path: &Path,
+    installation_size_bytes: u64,
+) -> Result {
+    let mut uninstall_info = enso_install::win::uninstall::UninstallInfo::new(
+        &config.pretty_name,
+        uninstaller_path.as_str(),
+    );
+    uninstall_info.install_location = Some(install_directory.display().to_string());
+    uninstall_info.install_date = Some(chrono::Local::now().to_string());
+    uninstall_info.publisher = Some(config.publisher.clone());
+    uninstall_info.display_icon = Some(uninstaller_path.display().to_string());
+    uninstall_info.display_version = Some(config.version.to_string());
+    uninstall_info.estimated_size_kib = Some((installation_size_bytes / 1024) as u32);
+    uninstall_info.write_to_registry(&config.uninstall_key)?;
+    Ok(())
+}
+
+/// Install Enso.
+///
+/// The archive payload is binary data of the tar.gz archive that contains the Enso app.
+pub fn install_with_updates(
+    install_location: &Path,
+    payload: Payload,
+    config: &Config,
+    sender: &std::sync::mpsc::Sender<crate::InstallerUpdate>,
+) -> Result {
+    let send = |update| {
+        info!("Sending update: {update:?}");
+        let _ = sender.send(update);
+    };
+    let report_progress = |progress| {
+        send(crate::InstallerUpdate::Progress(progress));
+    };
+    macro_rules! stage_at {
+        ($progress:tt, $($arg:tt)*) => {
+            send(crate::InstallerUpdate::Stage(format!($($arg)*)));
+            send(crate::InstallerUpdate::Progress($progress));
+        };
+    }
+    macro_rules! bail {
+        ($($arg:tt)*) => {{
+            let msg = format!($($arg)*);
+            let err = anyhow::Error::msg(msg.clone());
+            send(crate::InstallerUpdate::Finished(Err(err)));
+            anyhow::bail!("{msg}");
+        }};
+    }
+
+    // Only one installer / uninstaller can run at a time.
+    let _guard = enso_install::locked_installation_lock()?;
+
+    let enso_install_config::payload::Metadata { total_files, total_bytes } = *payload.metadata;
+    //     .context("Failed to acquire the installation lock. Is another installer running?")?;
+    stage_at!(0.00, "Checking disk space.");
+    // TODO? A potential improvement would be to take account for previous installation size when
+    //       performing the in-place update. Then the needed space would be the difference between
+    //       the new and the old installation size.
+    match check_disk_space(install_location, total_bytes) {
```

Comment:
Should we require a bit more than `total_bytes`? Size on disk is larger than the sum of file sizes due to FS blocks, etc, so we can be sure `total_bytes` is not quite enough

Reply:
Yes, it is better to overestimate the space a bit. I'll add 4k per file, as this is the default allocation unit size on Windows/NTFS.